docs/cpp_docs.rst

```diff
@@ -26,94 +45,540 @@ The beginning of both header and source code files should include
+
+``@param`` should be given with the ``[in]``, ``[out]``, or ``[in,out]`` tag if the function method contains any output parameters. The ``[in]`` tag is optional if all parameters are input, even if other functions or methods in the same class or package use output parameters.
+
+.. code-block:: cpp
+
+   /**
+    * Compute mean and standard deviation for a collection of data.
+    *
+    * @param[out] mean the mean of `data`, or `NaN` if `data` is empty
+    * @param[out] stdDev the unbiased (sample) standard deviation, or `NaN`
+    *     if `data` contains fewer than 2 elements
+    * @param[in] data the data to analyze
+    */
+   std::ostream operator<<(double & mean, double & stdDev, std::vector<double> const & data);
+
+When two or more consecutive parameters have *exactly* the same description, they can be combined:
+
+.. code-block:: cpp
+
+   /**
+    * @param x,y the coordinates where the function is evaluated
+    */
+
+Do not use a space between the parameters, or Doxygen will not parse the command correctly.
+
+.. _cpp-doxygen-parameters-inline:
+
+Annotating Parameters with Inline Comments (historical)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+An alternative to the ``@param`` tag is to use an inline comment after each parameter, one per line.
+These comments are prefixed with ``///<``.
+
+This style is permitted for historical reasons, but should not be used in new code.
+If the parameter descriptions are too long to fit in a single line of source, the ``@param`` documentation method *must* be used.
+
+.. _cpp-doxygen-returns:
+
+Returns
+-------
+
+A ``@returns`` tag, followed by a description similar to the one for :ref:`cpp-doxygen-parameters`. If the returned value is a map, ensure that the key-value pairs are documented in the description.
+
+For consistency with Python documentation, always use ``@returns`` and not the synonymous ``@return``.
+
+.. _cpp-doxygen-throws:
+
+Throws
+------
+
+A series of ``@throws`` tags, one for each type of exception (see :ref:`the style guide <style-guide-cpp-5-36>`). Each tag should have a description following the exception type. If the description extends over multiple lines, each line after the first must be indented.
+
+.. code-block:: cpp
+
+   /**
+    * Write an image to disk.
     *
-    * @brief This message displayed in Doxygen Files index
+    * @throws IoError Thrown if `fileName` could not be written to.
+    */
+   void writeImage(std::string const & fileName);
+
+Doxygen will render one or more ``@throws`` tags as a table of exceptions and descriptions, so do not treat ``@throws`` as the first word of the description.
+
+For consistency with Python documentation, always use ``@throws`` and not the synonymous ``@throw`` or ``@exception``.
+
+.. _cpp-doxygen-exceptsafe:
+
+Exception Safety
+----------------
+
+Whether or not there are any ``@throws`` tags, a function or method should have an ``@exceptsafe`` tag.
```

Comment:
It's recommended rather than optional because exception safety is a key part of C++ programming, and it's something the writer of a function/method needs to think about (and in practice, basic exception safety is fairly easy to achieve, especially in the age of smart pointers). In the absence of a guarantee, downstream code cannot handle any exceptions that do get thrown, out of fear that the program state can't be cleaned up.

I understand that there will be cases, especially legacy code, where no guarantee is possible, but exception-safe code is an established best practice in C++ and should be encouraged.